fix(blocks): retain data field from config on Update

mitchnielsen opened the pull request:
## Summary

When calling Update on a block, we call Get to retrieve the latest copy of the block object from the API. This response does not include the `$ref` expression if it was originally provided. This was causing "inconsistent result after apply" errors.

So for now, we'll use the user-defined data field when updating the state instead of the (incomplete) data field from the Get request.

Closes https://github.com/PrefectHQ/terraform-provider-prefect/issues/240

## Testing

Aside from the acceptance tests passing, I also validated manually to confirm the profile field showed the link to the profile as expected:

```terraform
resource "prefect_block" "my_dbt_cli_profile" {
  name      = "my-dbt-cli-profile"
  type_slug = "dbt-cli-profile"

  data = jsonencode({
    "name"   = "mine"
    "target" = "prefect-dbt-profile"
  })
}

resource "prefect_block" "my_dbt_run_operation_block" {
  name      = "my-dbt-operations"
  type_slug = "dbt-core-operation"

  # apply this first, without the `$ref` field (incorrect configuration)
  data = jsonencode({
    "commands"        = ["dbt deps", "dbt seed", "dbt run"]
    "dbt_cli_profile" = prefect_block.my_dbt_cli_profile.id
  })

  # apply this next, to trigger the Update logic (correct configuration)
  data = jsonencode({
    "commands"        = ["dbt deps", "dbt seed", "dbt run"]
    "dbt_cli_profile" = { "$ref" : { "block_document_id" : prefect_block.my_dbt_cli_profile.id } }
  })
}
```

<img width="550" alt="image" src="https://github.com/user-attachments/assets/70612d9a-b049-48cb-8f57-b0e76478d1ae">

mitchnielsen committed:
fix(blocks): retain data field from config on Update

When calling Update on a block, we call Get to retrieve the latest
copy of the block object from the API. This response does not include
the `$ref` expression if it was originally provided. This was causing
"inconsistent result after apply" errors.

So for now, we'll use the user-defined data field when updating the
state instead of the (incomplete) data field from the Get request.

Closes https://github.com/PrefectHQ/terraform-provider-prefect/issues/240

mitchnielsen committed:
Add example of "$ref" key usage

Adds an example of how the "$ref" key should be used.

github-actions[bot] committed:
Generate Terraform Docs

mitchnielsen commented on examples/resources/prefect_block/resource.tf:
```diff
@@ -47,3 +47,25 @@ resource "prefect_block" "gcp_credentials_key" {
 }
+
+# example:
+# some resources need to be referenced using a "$ref" key
```
I'd like to be more specific here so users know when the `$ref` key is needed. Looking into it.

mitchnielsen replied:
References to the application looking for this key:
- https://github.com/PrefectHQ/prefect/blob/2f000f9db4bef6bfc838e3a150168a4c48446045/src/prefect/utilities/templating.py#L246
- https://github.com/PrefectHQ/prefect/blob/2f000f9db4bef6bfc838e3a150168a4c48446045/src/prefect/server/models/block_documents.py#L116
- https://github.com/PrefectHQ/prefect/blob/2f000f9db4bef6bfc838e3a150168a4c48446045/src/prefect/server/api/validation.py#L115-L117
- https://github.com/PrefectHQ/prefect/blob/2f000f9db4bef6bfc838e3a150168a4c48446045/src/prefect/blocks/core.py#L95

mitchnielsen replied:
I was hoping it might be visible in the API docs, such as in [Create block document](https://app.prefect.cloud/api/docs#tag/Block-documents/operation/create_block_document_api_accounts__account_id__workspaces__workspace_id__block_documents__post) - but the `data` field is just described as `object (Data)` and the example payload is just `"data": {}`. So I'm looking to see how and end user would have known to format the data this way.

parkedwards replied:
yeah unfortunately i dont think we do a good enough job of documenting the JSON Schema part of the API, especially when it comes to Schema references like this ($ref)

https://json-schema.org/understanding-json-schema/structuring#dollarref

the best way I can illustrate when someone might need to make a reference like this is in the `prefect block type inspect` output:

```
➜ prefect block type inspect dbt-core-operation
```

in here, you'll notice most attributes with some `type`.  but if you look at `dbt_cli_profile`, it has the `$ref` expression:

![image](https://github.com/user-attachments/assets/a7f50b56-061f-4445-9d07-733774227e69)

for the most part, these references are for other Blocks (or rather, block_documents).  so we'd make a `$ref` to the `block_document_id`, which be a `prefect_block` resource's `.id` attribute

parkedwards replied:
a picture is worth a thousand words, so if you wanted to include a screenshot in the docs, i want to say we'd have to create a `template` for this page in the docs (kind of like our `templates/guides/getting-started.md`).  I'm a little rusty on the terraform-plugin-docs library when it comes to customizing `resource` doc pages, but i think you can create a custom templates for specific resource files

https://github.com/hashicorp/terraform-plugin-docs?tab=readme-ov-file#conventional-paths

parkedwards replied:
definitely not required, though

mitchnielsen replied:
```suggestion
# When referencing other Terraform resources from within the `data` attribute,
# the reference must be nested under a `$ref` key.
# docs: https://json-schema.org/understanding-json-schema/structuring#dollarref
```

Something like this may be more accurate/specific.

mitchnielsen replied:
Oops hadn't refreshed the page, checking your comments now.

mitchnielsen replied:
@parkedwards - nice, thanks for the pointers. Think I got it:

- [Rendered markdown doc](https://github.com/PrefectHQ/terraform-provider-prefect/blob/c28da120c52d06251976707b155dc47df32e07e9/docs/resources/block.md)
- [Image path replaced with PR branch name to confirm path is correct](https://raw.githubusercontent.com/PrefectHQ/terraform-provider-prefect/fix-block-data-update/docs/images/block-type-inspect-ref-key.png)

I created a template file for the Block document page, and added this new section into the template 👍🏼

parkedwards commented on internal/provider/resources/block.go:
```diff
@@ -418,13 +418,13 @@ func (r *BlockResource) Update(ctx context.Context, req resource.UpdateRequest,
+	// However, the API's GET method does not return the `$ref` expression if it
+	// was specified in the Data field on the Block resource. This leads to
+	// "inconsistent result after apply" errors. For now, we'll skip copying the
+	// retrieved Block's Data field and use what was specified in the plan.
```
👍

mitchnielsen committed:
Add doc template for Block resource

Adds the doc template for the Block resource to be rendered by
https://github.com/hashicorp/terraform-plugin-docs.

mitchnielsen committed:
Docs: add a note on when to use the dollarref
